### ISSUE TITLE
Shapes of intermediate tensors

### ISSUE THREAD

**Issue by justinchuby** [onnx/onnx#4580]
# Ask a Question

### Question
In a onnx graph, I can see the tensor shapes for the inputs and outputs. Is there a way to know what shapes the intermediate tensors are? I consulted the shape inferencing document but it looks like there is only a c++ api.

### Further information
- Relevant Area: shape_inference

### Notes

Model exported from pytorch

```python
import torch
from torch import nn

class Upsampling(nn.Module):
    def __init__(self):
        super().__init__()
        self.upsample = nn.Upsample(size=[2,2], mode='nearest')

    def forward(self, input, weight):
        weight =  self.upsample(weight)
        return torch.nn.functional.conv2d(input, weight)

torch.onnx.export(Upsampling(), (torch.tensor([[[[1,2],[1,2]]]], dtype=torch.float32), torch.tensor([[[[1]]]], dtype=torch.float32)), "test.onnx")
```

**Comment by jcwchen** [comment]
Not sure I understand this question correctly -- If you are looking for shape information of intermediate tensors after onnx.shape_inference, you could find them in `onnx_model.graph.value_info`.

**Comment by justinchuby** [comment]
Thanks! That's very helpful. For some reason the pytorch exported model does not include shape information. I needed to run `onnx.infer_shape` to obtain a model with non empty `value_info`

**Comment by justinchuby** [comment]
For this model

<img src="https://user-images.githubusercontent.com/11205048/193432597-4d9367ed-6719-4e96-b25a-3da31e8a23e4.png" width="200">

I am looking for the shape of `/upsample/Resize_output_0`.


I got 

```
[name: "/upsample/Shape_output_0"
type {
  tensor_type {
    elem_type: 7
    shape {
      dim {
        dim_value: 4
      }
    }
  }
}
, name: "/upsample/Constant_output_0"
type {
  tensor_type {
    elem_type: 7
    shape {
      dim {
        dim_value: 1
      }
    }
  }
}
, name: "/upsample/Constant_1_output_0"
type {
  tensor_type {
    elem_type: 7
    shape {
      dim {
        dim_value: 1
      }
    }
  }
}
, name: "/upsample/Constant_2_output_0"
type {
  tensor_type {
    elem_type: 7
    shape {
      dim {
        dim_value: 1
      }
    }
  }
}
, name: "/upsample/Slice_output_0"
type {
  tensor_type {
    elem_type: 7
    shape {
      dim {
        dim_value: 2
      }
    }
  }
}
, name: "/upsample/Constant_3_output_0"
type {
  tensor_type {
    elem_type: 7
    shape {
      dim {
        dim_value: 2
      }
    }
  }
}
, name: "/upsample/Concat_output_0"
type {
  tensor_type {
    elem_type: 7
    shape {
      dim {
        dim_value: 4
      }
    }
  }
}
, name: "/upsample/Resize_output_0"
type {
  tensor_type {
    elem_type: 1
    shape {
      dim {
        dim_param: "unk__0"
      }
      dim {
        dim_param: "unk__1"
      }
      dim {
        dim_param: "unk__2"
      }
      dim {
        dim_param: "unk__3"
      }
    }
  }
}
]
```

Do you have an idea why the dim may be unknown?

**Comment by jcwchen** [comment]
Seems like a typical case for ONNX data propagation since the shape information are computed dynamically. Shape, Slice, Concat are all supported for sure. I am not sure about Resize. Have you tried to enable data_prop in onnx_shape_inference? Please note that ONNX data propagation only supports opset_version>=13 for now.

**Comment by justinchuby** [comment]
Yes - both the inputs have a known shape

[test.onnx.zip](https://github.com/onnx/onnx/files/9748286/test.onnx.zip)

**Comment by justinchuby** [comment]
The input size looks to be `1x1x1x1` - was that what you were referring to or was it something else?

**Comment by jcwchen** [comment]
It needs the exact "tensor value" of input "sizes" for Resize, which comes from model.graph.input. The shape for the input "sizes" is not sufficient.

**Comment by jcwchen** [comment]
After offline discussion, we found that actually Resize is missing data propagation support so it cannot get shape information from previous Shape op. Then, I made a quick experiment and adding `ctx.getSymbolicInput()` in Resize op can indeed make it infer the right info. I will submit a PR for it when I have time. Thank you @justinchuby for catching this!
